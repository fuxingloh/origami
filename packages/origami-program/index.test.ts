import { join } from 'node:path';

import { expect, it } from '@jest/globals';

import { Compiler } from './index';

const compiler = new Compiler();

it('should compile program', async () => {
  const file = join(__dirname, 'tests', 'program.origami.ts');
  const program = await compiler.compile(file);

  expect(program).toEqual({
    version: 'v1',
    subscribe: [
      {
        from: 0,
        usi: 'eip155:1/block',
      },
    ],
    code: 'function h(h){return{hash:h.hash}}export{h as map};',
  });
});

it.each([
  'baseline.origami.ts',
  'async.origami.ts',
  'return-array.origami.ts',
  'multi-chain.origami.ts',
  'conditional.origami.ts',
  'erc20.origami.ts',
  'import-abi.origami.ts',
  'terser.origami.ts',
  'default-export.origami.ts',
])('should compile %s', async (file) => {
  const inputFile = join(__dirname, 'tests', file);
  const program = await compiler.compile(inputFile);
  expect(program).toMatchSnapshot();
});

it.each(['openzeppelin.origami.ts', 'viem.origami.ts', 'viem-large.origami.ts'])(
  'should compile with external dependency %s',
  async (file) => {
    const inputFile = join(__dirname, 'tests', file);
    const program = await compiler.compile(inputFile);

    // External dependency cause the code to change, hence we don't check them against a snapshot
    expect(program).toEqual({
      version: 'v1',
      subscribe: expect.any(Array),
      code: expect.any(String),
    });
  },
);

it.each([
  {
    file: 'fail.import-path.origami.ts',
    error: {
      message: expect.stringContaining("Unable to resolve import: 'path' in "),
    },
  },
  {
    file: 'fail.import-node-path.origami.ts',
    error: {
      message: expect.stringContaining("Unable to resolve import: 'node:path' in "),
    },
  },
  {
    file: 'fail.usi.origami.ts',
    error: {
      issues: [
        expect.objectContaining({
          code: 'invalid_string',
          path: [0, 'usi'],
        }),
      ],
    },
  },
  {
    file: 'fail.from.origami.ts',
    error: {
      issues: expect.arrayContaining([
        expect.objectContaining({
          code: 'invalid_type',
          path: [0, 'from'],
        }),
      ]),
    },
  },
])('should fail to compile $file', async ({ file, error }) => {
  const inputFile = join(__dirname, 'tests', file);
  await expect(compiler.compile(inputFile)).rejects.toMatchObject(error);
});
