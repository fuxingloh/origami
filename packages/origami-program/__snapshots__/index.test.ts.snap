// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should compile async.origami.ts 1`] = `
{
  "code": "async function h(h){return{hash:h.hash}}export{h as map};",
  "subscribe": [
    {
      "from": 0,
      "usi": "eip155:1/block",
    },
  ],
  "version": "v1",
}
`;

exports[`should compile baseline.origami.ts 1`] = `
{
  "code": "function h(h){return{hash:h.hash}}export{h as map};",
  "subscribe": [
    {
      "from": 0,
      "usi": "eip155:31337/block",
    },
  ],
  "version": "v1",
}
`;

exports[`should compile conditional.origami.ts 1`] = `
{
  "code": "function n(n){if(100n!==n.number)return 200n===n.number?{number:200n}:{number:n.number}}export{n as map};",
  "subscribe": [
    {
      "from": 0,
      "usi": "eip155:31337/block",
    },
  ],
  "version": "v1",
}
`;

exports[`should compile default-export.origami.ts 1`] = `
{
  "code": "function h(h){return{hash:h.hash}}export{h as map};",
  "subscribe": [
    {
      "from": 0,
      "usi": "eip155:1/block",
    },
  ],
  "version": "v1",
}
`;

exports[`should compile erc20.origami.ts 1`] = `
{
  "code": "function a(a){return{hash:a.transactionHash,block:{hash:a.blockHash,number:a.blockNumber},from:a.args.from,to:a.args.to,quantity:a.args.value}}export{a as map};",
  "subscribe": [
    {
      "abi": "erc20",
      "address": "0x6b175474e89094c44da98b954eedeac495271d0f",
      "from": 0,
      "to": 100,
      "usi": "eip155:31337/log",
    },
    {
      "abi": "erc20",
      "address": "0x6b175474e89094c44da98b954eedeac495271d0f",
      "from": 101,
      "usi": "eip155:31337/log",
    },
  ],
  "version": "v1",
}
`;

exports[`should compile import-abi.origami.ts 1`] = `
{
  "code": "function a(a){return{hash:a.transactionHash,block:{hash:a.blockHash,number:a.blockNumber}}}export{a as map};",
  "subscribe": [
    {
      "abi": [
        {
          "inputs": [
            {
              "internalType": "address",
              "name": "sender",
              "type": "address",
            },
            {
              "internalType": "uint256",
              "name": "tokenId",
              "type": "uint256",
            },
            {
              "internalType": "address",
              "name": "owner",
              "type": "address",
            },
          ],
          "name": "ERC721IncorrectOwner",
          "type": "error",
        },
        {
          "inputs": [
            {
              "internalType": "address",
              "name": "operator",
              "type": "address",
            },
            {
              "internalType": "uint256",
              "name": "tokenId",
              "type": "uint256",
            },
          ],
          "name": "ERC721InsufficientApproval",
          "type": "error",
        },
        {
          "inputs": [
            {
              "internalType": "address",
              "name": "approver",
              "type": "address",
            },
          ],
          "name": "ERC721InvalidApprover",
          "type": "error",
        },
        {
          "inputs": [
            {
              "internalType": "address",
              "name": "operator",
              "type": "address",
            },
          ],
          "name": "ERC721InvalidOperator",
          "type": "error",
        },
        {
          "inputs": [
            {
              "internalType": "address",
              "name": "owner",
              "type": "address",
            },
          ],
          "name": "ERC721InvalidOwner",
          "type": "error",
        },
        {
          "inputs": [
            {
              "internalType": "address",
              "name": "receiver",
              "type": "address",
            },
          ],
          "name": "ERC721InvalidReceiver",
          "type": "error",
        },
        {
          "inputs": [
            {
              "internalType": "address",
              "name": "sender",
              "type": "address",
            },
          ],
          "name": "ERC721InvalidSender",
          "type": "error",
        },
        {
          "inputs": [
            {
              "internalType": "uint256",
              "name": "tokenId",
              "type": "uint256",
            },
          ],
          "name": "ERC721NonexistentToken",
          "type": "error",
        },
        {
          "anonymous": false,
          "inputs": [
            {
              "indexed": true,
              "internalType": "address",
              "name": "owner",
              "type": "address",
            },
            {
              "indexed": true,
              "internalType": "address",
              "name": "approved",
              "type": "address",
            },
            {
              "indexed": true,
              "internalType": "uint256",
              "name": "tokenId",
              "type": "uint256",
            },
          ],
          "name": "Approval",
          "type": "event",
        },
        {
          "anonymous": false,
          "inputs": [
            {
              "indexed": true,
              "internalType": "address",
              "name": "owner",
              "type": "address",
            },
            {
              "indexed": true,
              "internalType": "address",
              "name": "operator",
              "type": "address",
            },
            {
              "indexed": false,
              "internalType": "bool",
              "name": "approved",
              "type": "bool",
            },
          ],
          "name": "ApprovalForAll",
          "type": "event",
        },
        {
          "anonymous": false,
          "inputs": [
            {
              "indexed": true,
              "internalType": "address",
              "name": "from",
              "type": "address",
            },
            {
              "indexed": true,
              "internalType": "address",
              "name": "to",
              "type": "address",
            },
            {
              "indexed": true,
              "internalType": "uint256",
              "name": "tokenId",
              "type": "uint256",
            },
          ],
          "name": "Transfer",
          "type": "event",
        },
        {
          "inputs": [
            {
              "internalType": "address",
              "name": "to",
              "type": "address",
            },
            {
              "internalType": "uint256",
              "name": "tokenId",
              "type": "uint256",
            },
          ],
          "name": "approve",
          "outputs": [],
          "stateMutability": "nonpayable",
          "type": "function",
        },
        {
          "inputs": [
            {
              "internalType": "address",
              "name": "owner",
              "type": "address",
            },
          ],
          "name": "balanceOf",
          "outputs": [
            {
              "internalType": "uint256",
              "name": "",
              "type": "uint256",
            },
          ],
          "stateMutability": "view",
          "type": "function",
        },
        {
          "inputs": [
            {
              "internalType": "uint256",
              "name": "tokenId",
              "type": "uint256",
            },
          ],
          "name": "getApproved",
          "outputs": [
            {
              "internalType": "address",
              "name": "",
              "type": "address",
            },
          ],
          "stateMutability": "view",
          "type": "function",
        },
        {
          "inputs": [
            {
              "internalType": "address",
              "name": "owner",
              "type": "address",
            },
            {
              "internalType": "address",
              "name": "operator",
              "type": "address",
            },
          ],
          "name": "isApprovedForAll",
          "outputs": [
            {
              "internalType": "bool",
              "name": "",
              "type": "bool",
            },
          ],
          "stateMutability": "view",
          "type": "function",
        },
        {
          "inputs": [],
          "name": "name",
          "outputs": [
            {
              "internalType": "string",
              "name": "",
              "type": "string",
            },
          ],
          "stateMutability": "view",
          "type": "function",
        },
        {
          "inputs": [
            {
              "internalType": "uint256",
              "name": "tokenId",
              "type": "uint256",
            },
          ],
          "name": "ownerOf",
          "outputs": [
            {
              "internalType": "address",
              "name": "",
              "type": "address",
            },
          ],
          "stateMutability": "view",
          "type": "function",
        },
        {
          "inputs": [
            {
              "internalType": "address",
              "name": "from",
              "type": "address",
            },
            {
              "internalType": "address",
              "name": "to",
              "type": "address",
            },
            {
              "internalType": "uint256",
              "name": "tokenId",
              "type": "uint256",
            },
          ],
          "name": "safeTransferFrom",
          "outputs": [],
          "stateMutability": "nonpayable",
          "type": "function",
        },
        {
          "inputs": [
            {
              "internalType": "address",
              "name": "from",
              "type": "address",
            },
            {
              "internalType": "address",
              "name": "to",
              "type": "address",
            },
            {
              "internalType": "uint256",
              "name": "tokenId",
              "type": "uint256",
            },
            {
              "internalType": "bytes",
              "name": "data",
              "type": "bytes",
            },
          ],
          "name": "safeTransferFrom",
          "outputs": [],
          "stateMutability": "nonpayable",
          "type": "function",
        },
        {
          "inputs": [
            {
              "internalType": "address",
              "name": "operator",
              "type": "address",
            },
            {
              "internalType": "bool",
              "name": "approved",
              "type": "bool",
            },
          ],
          "name": "setApprovalForAll",
          "outputs": [],
          "stateMutability": "nonpayable",
          "type": "function",
        },
        {
          "inputs": [
            {
              "internalType": "bytes4",
              "name": "interfaceId",
              "type": "bytes4",
            },
          ],
          "name": "supportsInterface",
          "outputs": [
            {
              "internalType": "bool",
              "name": "",
              "type": "bool",
            },
          ],
          "stateMutability": "view",
          "type": "function",
        },
        {
          "inputs": [],
          "name": "symbol",
          "outputs": [
            {
              "internalType": "string",
              "name": "",
              "type": "string",
            },
          ],
          "stateMutability": "view",
          "type": "function",
        },
        {
          "inputs": [
            {
              "internalType": "uint256",
              "name": "tokenId",
              "type": "uint256",
            },
          ],
          "name": "tokenURI",
          "outputs": [
            {
              "internalType": "string",
              "name": "",
              "type": "string",
            },
          ],
          "stateMutability": "view",
          "type": "function",
        },
        {
          "inputs": [
            {
              "internalType": "address",
              "name": "from",
              "type": "address",
            },
            {
              "internalType": "address",
              "name": "to",
              "type": "address",
            },
            {
              "internalType": "uint256",
              "name": "tokenId",
              "type": "uint256",
            },
          ],
          "name": "transferFrom",
          "outputs": [],
          "stateMutability": "nonpayable",
          "type": "function",
        },
      ],
      "from": 0,
      "usi": "eip155:31337/log",
    },
  ],
  "version": "v1",
}
`;

exports[`should compile multi-chain.origami.ts 1`] = `
{
  "code": "function e(e,t){if("eip155:31337/block"===t.usi){const t=e;return{type:"ethereum",hash:t.hash,number:t.number}}if("bip122:0f9188f13cb7b2c71f2a335e3a4fc328/block"===t.usi){const t=e;return{type:"bitcoin",hash:t.hash,number:BigInt(t.height)}}}export{e as map};",
  "subscribe": [
    {
      "from": 0,
      "usi": "eip155:31337/block",
    },
    {
      "from": 0,
      "usi": "bip122:0f9188f13cb7b2c71f2a335e3a4fc328/block",
    },
  ],
  "version": "v1",
}
`;

exports[`should compile return-array.origami.ts 1`] = `
{
  "code": "function t(t){return[{type:"block",height:BigInt(t.height)},{type:"tx",height:BigInt(t.height)}]}export{t as map};",
  "subscribe": [
    {
      "from": 0,
      "usi": "bip122:0f9188f13cb7b2c71f2a335e3a4fc328/block",
    },
  ],
  "version": "v1",
}
`;

exports[`should compile terser.origami.ts 1`] = `
{
  "code": "function n(n){return{number:n.number,a:"a",kb:102400}}export{n as map};",
  "subscribe": [
    {
      "from": 0,
      "usi": "eip155:31337/block",
    },
  ],
  "version": "v1",
}
`;
